programs/marinade-finance/src/instructions/management/redelegate.rs

```diff
@@ -0,0 +1,394 @@
+        self.state.stake_system.add(
+            &mut self.stake_list.data.as_ref().borrow_mut(),
+            &self.redelegate_stake_account.key(),
+            redelegate_amount_effective,
+            &self.clock,
+            0, // is_emergency_unstaking
+        )?;
+
+        // we now consider amount no longer "active" for this specific validator
+        source_validator.active_balance -= redelegate_amount_effective;
+        // it moved to dest-validator
+        dest_validator.active_balance += redelegate_amount_effective;
+
+        // update stake-list & validator-list
+        self.state.stake_system.set(
+            &mut self.stake_list.data.as_ref().borrow_mut(),
+            stake_index,
+            stake,
+        )?;
+        self.state.validator_system.set(
+            &mut self.validator_list.data.as_ref().borrow_mut(),
+            source_validator_index,
+            source_validator,
+        )?;
+        self.state.validator_system.set(
+            &mut self.validator_list.data.as_ref().borrow_mut(),
+            dest_validator_index,
+            dest_validator,
+        )?;
+
+        Ok(())
+    }
+
+    pub fn return_rent_unused_stake_account(
+        &self,
+        unused_stake_account: AccountInfo<'info>,
+    ) -> Result<()> {
+        // Return back the rent reserve of unused stake account (split or redelegate reserve)
+        withdraw(
+            CpiContext::new(
+                self.stake_program.to_account_info(),
+                Withdraw {
+                    stake: unused_stake_account.clone(),
+                    withdrawer: unused_stake_account.clone(),
+                    to: self.split_stake_rent_payer.to_account_info(),
+                    clock: self.clock.to_account_info(),
+                    stake_history: self.stake_history.to_account_info(),
+                },
+            ),
+            unused_stake_account.lamports(),
+            None,
+        )
+    }
+
+    #[inline] // separated for readability
+    pub fn split_stake_for_redelegation(&mut self, mut stake: StakeRecord, amount: u64) -> Result<()> {
+        msg!(
+            "Split {} lamports from stake {} to {}",
+            amount,
+            stake.stake_account,
+            self.split_stake_account.key(),
+        );
+
+        // add the split account as new account to Marinade stake-accounts list
+        self.state.stake_system.add(
+            &mut self.stake_list.data.as_ref().borrow_mut(),
+            &self.split_stake_account.key(),
+            0, // this account will be deactivating,
+            // all lamports will be moved to the re-delegated account,
+            // but even with no lamports, we expect the redelegate-deactivating account to provide rewards at the end of the epoch.
+            // After completing deactivation, whatever is there minus rent is considered last rewards for the account
+            &self.clock,
+            1, // is_emergency_unstaking
```

Comment:
No emergency unstaking because it is 0-last_update_delegated_lamports stake record